docs/adrs/004-SplitAndAggregate.md

```diff
@@ -0,0 +1,198 @@
+
+### LIMIT KBP jobs to a maximum of N<10 active KBP jobs per user
+* `+` Simple solution requires ee2 to maintain list of KBP apps, and add a KBP_LIMIT to jobs from this list. [Condor](https://github.com/kbase/condor/pull/26) will need KBP_LIMIT Added
+* `+` List of apps is not frequently updated
+* `+` Apps do not need to be modified
+* `-` If a new app uses KBP and their app is not on the list, it won't be limited by the KBP_LIMIT unless the owner lets us know.
+* `-` If an existing app no longer uses KBP, their app is still limited unless the owner lets us know.
+* `-` Nodes will continue be taken up by jobs that do little computation as each job gets its own node.
+* `-` Users may not be able to effectively use up their 10 job spots 
+* `-` Capacity is still wasted, as some KBP jobs sit around waiting for other jobs to run
+
+###  LIMIT KBP jobs to a maximum of N<10 active jobs per user + Seperate queue for kbparallels apps 
+* `+` Same pros as above
+* `+` Allows us to group up KBP jobs onto fewer machines, instead of giving them their entire node
+* `-` Requires going through each app and understanding the worst case computational needs in order to  set the estimated cpu and memory needs for each app 
+* `-` Apps can interfere with other innocent apps and take them down
+* `-` Creating a new queue requires balancing between how many active KBP nodes there vs how many nodes are available for other NJS jobs.
+* `-` Capacity is still wasted, as some KBP jobs sit around waiting for other jobs to run
+
+
+### Build KBP Lightweight Version + KBP Queue
+
+
+#### Design of new verison
+* All apps must be modified to use the new KBP lightweight version, which will: 
+* Can either modify KBP, or create a new tool/package to use instead of KBP
+
+1) Launch a management job called the *Job Manager* that sits in the KBP Queue, alongside other KBP jobs. Other jobs are launched in the NJS queue.
+2) Launch the *Setup Job* which will
+ * Use the *User Parameters* and/or
+ * Optionally Download the Data from the *User Parameters*  to figure out *Job Manager* parameters
+ * Use the results of information gathered from the initial download and or *User Parameters*
+ * Generate final parameters to be sent to the *Job Manager* to launch *Fan Out* jobs
+3) The *Job Manager* now has enough parameters to setup *Fan Out* Jobs, and will send out jobs and wait for them (and possibly retry them upon failure)
+4) *Fan Out* jobs download data and perform calculations, save them back to the system, and return references to the saved objects
+5) The *Job Manager* optionally launches a *Group/Reduce* job based on User Parameters and or the results of *Fan Out* Jobs
+6) The *Group/Reduce* job downloads objects from the system, and creates a set or other grouping, and then saves the set back to the system
+7) The *Job Manager*  sends the results back to a *Report* Job, which downloads the final data and uploads a report based on that data
```

Comment:
I'm not sure why the report couldn't be created in the group/reduce job?

Reply:
What if the report needs to have a value that is only know once the group/reduce job has completed?

Reply:
But the group/reduce/FI job produces all the data / objects that are the end result of the job, so it has to know all the values before it ends execution, unless I'm missing something.

Reply:
Sounds like we need to do another call

Reply:
On second thought, if we send all the data from previous steps into the reduce job, it could do as you describe. However, if we don't have a reduce step, we'd still need a job to download the data and to create a report

Reply:
Is there an assumption that we are always outputting a set from a KBP job?

Reply:
> However, if we don't have a reduce step, we'd still need a job to download the data and to create a report

To me both of those are just FI jobs, whether it's reduce, report, or reduce and report. For some jobs you might not need a FI step at all.

> Is there an assumption that we are always outputting a set and a report from a KBP job?

I don't think so.

Reply:
:+1: